MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,339 @@
+  Learning rate schedule has either two or three parts:
+  1) Linear warmup from 0 to [learning_rate] over steps 0 to [learning_rate_schedule_steps * warmup_steps_fraction]
+  2) Cosine from [learning_rate] to [learning_rate * cosine_learning_rate_final_fraction] until learning_rate_schedule_steps
+  3) Constant learning rate of 0 from learning_rate_schedule_steps to steps.
+  The zero learning rate section can be used to more accurately measure the fully trained model's performance.
+  """
+  def make_cos_schedule(init_lr, final_lr, len_steps):
+    def schedule(step):
+      pct = (step) / len_steps
+      a = 0.5 * (jnp.cos(jnp.pi*pct) + 1)
+      lr = init_lr * a + final_lr * (1 - a)
+      return lr
+    return schedule
+
+  lr = config.learning_rate
+  cos_final_lr = lr * config.cosine_learning_rate_final_fraction
+
+  warmup_steps = int(config.learning_rate_schedule_steps * config.warmup_steps_fraction)
+  cos_steps = config.learning_rate_schedule_steps - warmup_steps
+  constant_zero_steps = config.steps - config.learning_rate_schedule_steps
+
+  warmup_schedule = optax.linear_schedule(
+      init_value=0.0,
+      end_value=lr,
+      transition_steps=warmup_steps
+  )
+  cos_schedule = make_cos_schedule(lr, cos_final_lr, cos_steps)
+  constant_schedule = optax.constant_schedule(0.0)
+
+  pieces = [warmup_schedule, cos_schedule]
+  boundaries=[
+   warmup_steps,
+   warmup_steps + cos_steps,
+   ]
+
+  if constant_zero_steps > 0:
+    pieces.append(constant_schedule)
+    boundaries.append(warmup_steps + cos_steps + constant_zero_steps)
+
+  return optax.join_schedules(pieces, boundaries)
+
+def get_abstract_state(model, tx, config, rng, mesh):
+  """ Get a shaped abstraction of the state (including optimizer)"""
+  init_train_state_partial = functools.partial(init_train_state, model, tx,
+                                               config)
+  abstract_state = jax.eval_shape(init_train_state_partial, rng)
+  state_logical_annotations = nn.get_partition_spec(abstract_state)
+  unboxed_abstract_state = unbox_logicallypartioned_trainstate(abstract_state)
+
+  # Initialization
+  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+    state_mesh_annotations = nn.logical_to_mesh(state_logical_annotations)
+  return unboxed_abstract_state, state_mesh_annotations
+
+def unbox_logicallypartioned_trainstate(
+    boxed_train_state: train_state.TrainState):
+  """ Unboxes the flax.LogicallyPartitioned pieces in a train state.
+
+    Args:
+      boxed_train_state: a train state that includes LogicallyPartitioned
+        leaves.
+    Returns:
+      a TrainState where all all LogicallyPartitioned leaves have been unboxed.
+  """
+  return jax.tree_util.tree_map(lambda x: x.unbox() if \
+        isinstance(x, flax.linen.spmd.LogicallyPartitioned) \
+        else x, boxed_train_state, \
+        is_leaf=lambda k: isinstance(k, flax.linen.spmd.LogicallyPartitioned))
+
+def load_compiled(config, partial_train, state):
```

Comment:
Currently this can only load the maxtext trainstep file because input shapes are required. Parker is working on removing this requirement in which case load_compiled will really just be deserailizing the the compiled and then it can be moved to the shared directory